lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java

```diff
@@ -276,33 +321,106 @@ public void putAll(Map<? extends K, ? extends V> m)
             @Override
             public void clear()
             {
-                delegate.clear();
+                dataCacheMap.clear();
+                tokens.clear();
             }
 
             @Override
             public Set<K> keySet()
             {
-                return delegate.keySet();
+                return tokens.keySet();
             }
 
             @Override
             public Collection<V> values()
             {
-                // values() should be a view, but also, it has a size and, iterating values shouldn't throw for incomplete futures.
-                throw new UnsupportedOperationException();
+                return dataCacheMap.values();
             }
 
             @Override
-            public Set<Entry<K, V>> entrySet()
+            public Set<Map.Entry<K, V>> entrySet()
             {
                 throw new UnsupportedOperationException();
             }
         };
     }
 
-    @Override
-    public void cleanUp()
+    // instance-based equality
+    static final class Token<K>
     {
-        delegate.cleanUp();
+        private final K key;
+
+        Token(K key)
+        {
+            this.key = requireNonNull(key, "key is null");
+        }
+
+        K getKey()
+        {
+            return key;
+        }
+
+        @Override
+        public String toString()
+        {
+            return format("CacheToken(%s; %s)", Integer.toHexString(hashCode()), key);
```

Comment:
From `hashCode` javadoc:
```
As much as is reasonably practical, the hashCode method defined
by class {@code Object} does return distinct integers for
distinct objects
```

IIUC this says same hashcode can be returned for different objects present in VM at the same time. 
Given that does it make sense to put hashcode as part of the `toString` result?

Reply:
i found it helpful and default `Object.toString` does exactly that, but i see it could be misleading.